### PR TITLE
[extension/encodingextension] Implement encoding extension

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Implement a first codec for the encoding extension.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28686

**Testing:**
Unit tests.

**Documentation:**
Some package docs.

### CODE REVIEW

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by atoulme** [comment]
Going to wait until after the release to take care of conflicts, more reviews are appreciated.

File: extension/encodingextension/config.go
Comment (dmitryax) [inline extension/encodingextension/config.go:13]: @MovieStoreGuy @atoulme 

This would require all the codecs to be added in this repo. I'm a bit worried about additional burden on the maintainers.

What if we have the encoding/decoding extensions as a type of extensions similar to the storage extension?

Each encoding/decoding extension would require to implement some API and carry the codec inside of it instead of exposing it as an additional public API?

In that case, we would have `jaegerencoder` and `jaegerdecoder` extensions and others implementing already existing `traces.Marshal` and `traces.Unmarshal`.

Users could implement their own encoding/decoding extension outside of this repo.

WDYT?

Comment (atoulme) [reply extension/encodingextension/config.go:13]: I need to review again how we work with the storage extension, I think that's a good idea in terms of UX.

Comment (MovieStoreGuy) [reply extension/encodingextension/config.go:13]: Makes sense to me, it would definitely unblock development and make it easier to build your own.

Comment (atoulme) [reply extension/encodingextension/config.go]: OK I have reimplemented with your suggestions in mind. PTAL

File: extension/encodingextension/codec.go
Comment (dmitryax) [inline extension/encodingextension/codec.go:1]: I don't think any API in this file is needed. We probably don't need encoding extension at all. We need extensions for particular codecs only.

Let's say kafka receiver uses another tracing encoder. The code would be something like this:

```golang
func (f *kafkaReceiverFactory) createTracesReceiver(
	_ context.Context,
	set receiver.CreateSettings,
	cfg component.Config,
	nextConsumer consumer.Traces,
) (receiver.Traces, error) {
	c := cfg.(*Config)
	ext, found := host.GetExtensions()[c.Decoder]
	!found {
		return nil, fmt.Errorf("extension %s not found", c.Decoder)
	}
	decoder, ok := ext.(traces.Unmarshaler)
	if !ok {
		return nil, fmt.Errorf("extension %s is not a traces decoder")
	}
	... decoder.UnmarshalTraces(...)
}

```

Comment (atoulme) [reply extension/encodingextension/codec.go:1]: we can drop more API as needed and just use the Marshaler/Unmarshaler types to make it simpler.

Comment (atoulme) [reply extension/encodingextension/codec.go:1]: Done.

Comment (dmitryax) [reply extension/encodingextension/codec.go:1]: Ok, in that case, we don't need the encodingextension. If you starting with jaeger, let's rename it to `jaegerencodingextension`

Comment (atoulme) [reply extension/encodingextension/codec.go:1]: Sure, I can cut down some more.

File: .github/CODEOWNERS
Comment (dmitryax) [inline .github/CODEOWNERS:96]: Why embedding? We don't follow this pattern in other components. I think it should be 

`extension/jaegerencodingextension` and `extension/textencodingextension`

2 more Q: 

1. Where do you want to use the text encoding extension?
2. Why are you not adding yourself as the codeowner? :)

Comment (atoulme) [reply .github/CODEOWNERS:96]: I'll remove the embedding. We do embed for storage but it's fine, we don't need to do that.

1. We use the `text` encoding for kafka/google pubsub/azure eventhub. We also use a bytes encoding. I know.
2. You really think I'm not going to jump at the opportunity to leave @MovieStoreGuy with the bag? If you insist I can add myself.

Comment (dmitryax) [reply .github/CODEOWNERS:96]: > I'll remove the embedding. We do embed for storage but it's fine, we don't need to do that.

Oh, you're right. I was wrong. 

What about `extension/encoding/jaegerencongingextension`? I think it's important to have the `extension` part in the module/package name. Not sure why storage extension doesn't do that

Comment (dmitryax) [reply .github/CODEOWNERS:96]: > You really think I'm not going to jump at the opportunity to leave @MovieStoreGuy with the bag? If you insist I can add myself.

Please add yourself. I thing @MovieStoreGuy would agree :)

Comment (atoulme) [reply .github/CODEOWNERS:96]: OK no worries. Thanks for the quick feedback. I should have this done by tonight, just shuffling things around.

Comment (atoulme) [reply .github/CODEOWNERS]: Pushed with your recommendations.

File: extension/encoding/jaegerencodingextension/config.go
Comment (dmitryax) [inline extension/encoding/jaegerencodingextension/config.go:12]: This should be exposed as enum as well

```suggestion
	JaegerProtocolProtobuf JaegerProtocol = "protobuf"
```

File: versions.yaml
Comment (dmitryax) [inline versions.yaml:111]: This is not needed, right?

Comment (atoulme) [reply versions.yaml:111]: ah, nope. Too many files flying around... I'll fix it.

**Comment by atoulme** [comment]
@MovieStoreGuy please take another look, things have changed a lot since you approved the first time around.

**Review by MovieStoreGuy — APPROVED** [review]
This is great! Thank you for pushing this forward!

**Pull request merged**
